api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,141 @@
+  map<string, Policy> policies = 2;
+}
+
+// Policy specifies a role and the principals that are assigned/denied the role.
+message Policy {
+ // Required. The set of permissions that define a role.
+ repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+
+ // Required. List of principals that are assigned/denied the role based on “action_type”.
+ repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+message StringMatch {
+  oneof match_pattern {
+    // The input string must match exactly the string specified here.
+    // Or it is a "*", which means that it matches any string.
+    string simple = 1;
+
+    // The input string must have the prefix specified here.
+    string prefix = 2;
+
+    // The input string must have the suffix specified here.
+    string suffix = 3;
+
+    // The input string must match the regular expression specified here.
+    // The regex grammar is defined `here
+    // <http://en.cppreference.com/w/cpp/regex/ecmascript>`_.
+    string regex = 4;
+ }
+}
+
+// Specifies how to match an entry in a map.
+message MapEntryMatch {
+  // The key to select an entry from the map.
+  string key = 1;
+
+  // A list of matched values.
+  repeated StringMatch values = 2;
+}
+
+// Permission defines a permission to access the service.
+message Permission {
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // If this field is unset, it applies to any path.
+  repeated StringMatch paths = 1;
+
+  // Required. A list of HTTP methods (e.g., "GET", "POST").
+  // If set to ["*"], it applies to any method.
+  repeated string methods = 2;
+
+  // Definition of a custom condition.
+  message Condition {
+    oneof condition_spec {
+      // Header match. This refers to the "request.http.headers" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      MapEntryMatch header = 1;
+
+      // Label match. This refers to the "destination.labels" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      MapEntryMatch label = 2;
+    }
+  }
+
+  // Optional. Custom conditions.
+  repeated Condition conditions = 3;
+}
+
+// Principal defines an identity or a group of identities.
+message Principal {
+  // Optional. The name of the principal. This refers to the "source.principal" field in
+  // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+  // If set to "*", it applies to any user.
+  string name = 1;
+
+  // Definition of a custom attribute to identify the principal.
+  message Attribute {
+    oneof attribute_spec {
+      // Header match. This refers to the "request.http.headers" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      MapEntryMatch header = 1;
```

Comment:
Why are we header matching within Principal and within Permissions?

Reply:
A policy says "who" can do "what". The Principal is the "who" part. And the Permission is the "what" part. Different headers can be used to describe either "who" or "what". For example, "x-api-key" header describes the Principal ("who": api-key). "x-service-version" header describes the Permission ("what": service version).